connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/ResponseValidator.scala

```diff
@@ -42,8 +51,24 @@ class ResponseValidator extends Logging {
     serverSideSessionId = Some(serverSideSessionId.getOrElse("") + suffix)
   }
 
+  /**
+   * Returns true if the server side session ID has changed.
+   */
+  private[sql] def hasSessionChanged: Boolean = {
+    sessionInvalidated.getAcquire
+  }
+
   def verifyResponse[RespT <: GeneratedMessageV3](fn: => RespT): RespT = {
-    val response = fn
+    val response =
+      try {
+        fn
+      } catch {
+        case e: StatusRuntimeException
+            if e.getStatus.getCode == Status.Code.INTERNAL &&
+              e.getMessage.contains("[INVALID_HANDLE.SESSION_CHANGED]") =>
+          sessionInvalidated.setRelease(true)
```

Comment:
Should we also reset the serverSideSessionId?

Reply:
I don't think so because serverSideSessionId will be updated to the new value if we reset it here. Maybe @nemanja-boric-databricks wants to double check this if I missed something here.